esteid/actions.py

```diff
@@ -119,227 +122,429 @@ def do_action(cls, view: "GenericDigitalSignViewMixin", params: dict = None, *,
+
+        # now we don't need the session anymore
+        delete_esteid_session(request)
 
-        if temp_container_file:
-            # Load a partially prepared BDoc from a tempfile and clean it up
-            container = BDoc2File(temp_container_file)
-            os.remove(temp_container_file)
-        else:
-            container = BDoc2File(container_path)
+        verify(xml_sig.get_certificate_value(), signature_value, binascii.a2b_base64(signed_hash_b64), prehashed=True)
 
-        xml_sig.add_signature_value(signature_value)
+        xml_sig.set_signature_value(signature_value)
+        issuer_cert = get_certificate(xml_sig.get_certificate_issuer_common_name())
 
         try:
-            finalize_signature(xml_sig)
-        except BDoc2Error:
+            finalize_signature(xml_sig, issuer_cert, lt_ts=ESTEID_USE_LT_TS, ocsp_url=OCSP_URL, tsa_url=TSA_URL)
+        except pyasice.Error:
             logger.exception("Signature confirmation service error")
             return {
-                'success': False,
-                'code': 'SERVICE_ERROR',
+                "success": False,
+                "code": "SERVICE_ERROR",
             }
 
         container.add_signature(xml_sig)
 
         return {
-            'success': True,
-            'container': container,
+            "success": True,
+            "container": container,
         }
 
 
 class SignCompleteAction(BaseAction):
     @classmethod
     def do_action(cls, view, action_kwargs):
-        service = view.get_service()
-
-        return service.get_file_data(view.get_files())
+        """Return the signed container"""
+        raise NotImplementedError
 
 
 class MobileIdSignAction(BaseAction):
     @classmethod
-    def do_action(cls, view, action_kwargs):
-        service = view.get_service()
+    def do_action(
+        cls,
+        view: "GenericDigitalSignViewMixin",
+        params: dict = None,
+        *,
+        phone_number: str = None,
+        id_code: str = None,
+        language: str = None,
+    ):
+        """
+        The old API is to pass a dict of params (previously confusingly named `action_kwargs`),
+        the keyword args are added here for clarity as to what the method accepts
+
+        :param view:
+        :param params:
+        :param phone_number:
+        :param id_code:
+        :param language:
+        :return:
+        """
+        request = view.request
+        delete_esteid_session(request)
 
-        if not view.get_files():
+        if not phone_number:
+            phone_number = params["phone_number"]
+        if not id_code:
+            id_code = params["id_code"]
+
+        if not (phone_number and id_code):
             return {
-                'success': False,
-                'code': 'MIN_1_FILE',
+                "success": False,
+                "code": "BAD_PARAMS",
             }
 
-        # Create signed document
-        service.create_signed_document()
+        if not re.match(r"\d{11}$", id_code):
```

Comment:
Not sure if this is a good idea to validate this as all.

On one hand, probably nothing bad happens if we allow any number of digits or even arbitrary text (i.e. the API where this will be passed will fail still if ID code is invalid), so for sake of not assuming things it makes sense to let the API to handle this validation and errors.

On other hand, it is likely that for quite a long time the requirement of ID code still being 11 digits will hold, probably for longer than life cycle of this library.

Reply:
The point of validating it is to avoid a request to upstream, which would also most likely result in a very generic "Invalid parameters" and assuming user has just mistyped their code, it would be quicker and friendlier to just say that ASAP. If it were a malicious user, then sending arbitrary strings to upstream services may also lead to more serious issues. In fact here should be a proper validation which is no rocket science in itself but I just didn't plan this actions api to last long (that's why this new API proposal) so just glossed over it.